contracts/token/ERC721/extensions/ERC721Wrapper.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../ERC721.sol";
+import "../utils/ERC721Holder.sol";
+
+/**
+ * @dev Extension of the ERC721 token contract to support token wrapping.
+ *
+ * Users can deposit and withdraw an "underlying token" and receive a "wrapped token" with a matching tokenId. This is useful
+ * in conjunction with other modules. For example, combining this wrapping mechanism with {ERC721Votes} will allow the
+ * wrapping of an existing "basic" ERC721 into a governance token.
+ *
+ * _Available since v4.9.0_
+ */
+abstract contract ERC721Wrapper is ERC721, ERC721Holder {
+    IERC721 public immutable underlying;
+
+    constructor(IERC721 underlyingToken) {
+        underlying = underlyingToken;
+    }
+
+    /**
+     * @dev Allow a user to deposit underlying tokens and mint the corresponding tokenIds.
+     */
+    function depositFor(address account, uint256[] memory tokenIds) public virtual returns (bool) {
+        bytes memory data = abi.encode(account);
+
+        uint256 length = tokenIds.length;
+        for (uint256 i = 0; i < length; ++i) {
+            underlying.safeTransferFrom(_msgSender(), address(this), tokenIds[i], data);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev Allow a user to burn wrapped tokens and withdraw the corresponding tokenIds of the underlying tokens.
+     */
+    function withdrawTo(address account, uint256[] memory tokenIds) public virtual returns (bool) {
+        uint256 length = tokenIds.length;
+        for (uint256 i = 0; i < length; ++i) {
+            uint256 tokenId = tokenIds[i];
+            require(_isApprovedOrOwner(_msgSender(), tokenId), "ERC721Wrapper: caller is not token owner or approved");
+            _burn(tokenId);
+            // Checks were already performed at this point, and there's no way to retake ownership or approval from
+            // the wrapped tokenId after this point, so it's safe to remove the reentrancy check for the next line.
+            // slither-disable-next-line reentrancy-no-eth
+            underlying.safeTransferFrom(address(this), account, tokenId);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev Overrides {IERC721Receiver-onERC721Received} to allow minting on direct ERC721 transfers to
+     * this contract.
+     *
+     * WARNING: Doesn't work with unsafe transfers (eg. {IERC721-transferFrom}). Use {ERC721Wrapper-_recover}
+     * for recovering in that scenario.
+     */
+    function onERC721Received(
+        address,
+        address from,
+        uint256 tokenId,
+        bytes memory data
+    ) public override returns (bytes4) {
+        require(msg.sender == address(underlying));
+        _safeMint(data.length == 0 ? from : abi.decode(data, (address)), tokenId);
```

Comment:
This seems risky to me. How do we know the transfer initiator isn't just passing some other data in here for whatever reason?

Instead of assuming any non-empty data encodes a receiver address, I would prefer if we required some magic value to be present, like a 4 byte selector, at the beginning of the data.

Reply:
Would like to hear @Amxx's thoughts on this.

Reply:
> How do we know the transfer initiator isn't just passing some other data in here for whatever reason?

The `underlying` is the only one who can call `onERC721Received`, so whoever is deploying this should make sure the underlying pass the correct data and we can't enforce. So I'd say the `underlying` is trusted. [The data is encoded in the `depositFor` function.](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3863/files#diff-0b3d65cc80b6f6aa8b2eec2949c36d3eba9035fa132d90d367cc62291d97dafdR28)

> I would prefer if we required some magic value to be present, like a 4 byte selector, at the beginning of the data.

We can append ERC721Wrapper's `interfaceId` with a value like:

```solidity
    bytes32 private constant WRAPPER_AWARE = type(ERC721Wrapper).interfaceId;
```

and use it in depositFor as:

```solidity
        bytes memory data = abi.encode(WRAPPER_AWARE, account);
```

Then just check the prefix in `onERC721Check`.

Also, I made sure they produce different interfaceIds:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

interface A {
    function interfaceId() external pure returns(bytes4);
}

abstract contract B is A {
    function interfaceId() external pure override returns(bytes4) {
        return type(A).interfaceId;
    }

    function anotherInterfaceId() public pure returns(bytes4) {
        return type(B).interfaceId;
    }
}

contract C is B {}
```

To be honest, I think it's too much, but will wait for comments from @Amxx

Reply:
>so whoever is deploying this should make sure the underlying pass the correct data

This is not something anyone can guarantee. ERC721 allows any token holder to initiate a transfer with arbitrary data attached to it, to be passed through to `onERC721Received`. Just as ERC721Wrapper is encoding the address in `depositFor`, a wallet may decide to include anything in there too. It would be slightly reckless to do so because it could have consequences, but I wouldn't find it too surprising to learn of some wallet that includes data to track wallet use or something.

Reply:
> We can append ERC721Wrapper's `interfaceId` with a value like:

This sounds good if we go forward with the magic value. Note that it would be more efficient to use `abi.encodePacked` rather than `abi.encode`.

Reply:
> I wouldn't find it too surprising to learn of some wallet that includes data to track wallet use or something.

This is a good example of a use case. Thanks!

> This sounds good if we go forward with the magic value.

I realized it may not be backwards compatible if we introduce a new function to the interface, so I decided to use a hardcoded bytes4 value.

4 bytes are arbitrary, though. We can use the last 12 bytes to pack it with the address. Keeping it 4 bytes will keep the risk of a collision in 1/2**32.

[Here's the implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3863/commits/fb22e8dfec224f49bb6730846e3a3dd929332a73)

> abi.encodePacked is even more efficient but in that case I don't think we can use abi.decode, this may be the best choice though.

I used `abi.encodePacked` with a decoding private function. Why would you keep `abi.encodeWithSelector`?

Reply:
In your implementation you require the magic value to be present even when the token should be minted to the sender. I don't think this is necessary. It's safe to assume that the sender wants to wrap the token and receive it in the same account if data is empty.

Reply:
[Updated and simplified.](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3863/commits/24b103306346c8511b8ea38b1047888aa0ae8c51)